### PR TITLE
[Kernel] W8A16 Int8 inside FusedMoE 

### PR DESCRIPTION
**Pull request by mzusman**
### 🚀  The feature, motivation and pitch

Additional feature for `fused_moe` triton kernel to support W8A16 with Int8, supports Ampere/Ada lovelace/Hopper, called ExpertsInt8.
Based on symmetric per-column per-expert Int8 quantization, casting the weights to FP16/BF16 before matmul inside the `fused_moe` kernel (compute_type in FP/BF16).
Support quantization and scales extraction on startup (takes 1min on Jamba). 
We've ran quality benchmarks on [Jamba](https://huggingface.co/ai21labs/Jamba-v0.1) and it shows no quality degradation:

Method | BF16 | ExpertsInt8
-- | -- | --
gsm8k cot | 59.5 | 59.8
MMLU | 67.3 | 67.3
gsm8k | 50.6 | 50.1
narrative_qa | 67.8 | 68.7
ppl/c4 | -0.4301 | -0.4301

Performance: 
E2E latency in seconds on requests with Prompt length=1024, decode length=128:
Model | Hardware | Method | BS=1 | BS=4 | BS=8
-- | -- | -- | -- | -- | --
Mixtral8x22B | H100*8 | FP8 (MoE quant-only) | 1.3 | 1.79 | 2.26
Mixtral8x22B | H100*8 | ExpertsInt8 | 1.3 | 1.84 | **2.205**
-- | -- | -- | -- | -- | --
Mixtral8x7B | H100*4 | FP8 (MoE quant-only) | 0.835 | 1.22 | 1.44
Mixtral8x7B | H100*4 | ExpertsInt8 | 0.83 | 1.22 | **1.42**
-- | -- | -- | -- | -- | --
Jamba | H100*2 | FP8 (MoE quant-only) | 1.3 | 2 | 3.2
Jamba | H100*2 | ExpertsInt8 | **1.16** | **2** | **3.14**
-- | -- | -- | -- | -- | --
Jamba | A100*2 | FP16 | 1.75 | 3.2 | 4.6
Jamba | A100*2 | ExpertsInt8 | **1.65** | **2.7** | **4**
Jamba | A100*2 | GPTQ 8bit (W/o FusedMoe) | 3.8 | 5.9 | 9.4

### Advantages: 
* Doesn't require calibration preprocess.
* No quality degradation.
* Quantized FusedMoE methodology that runs on A100s.
* Safer in case of large activations since they can be saved in BF16 reducing the risk of overflow.

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
Hey - just FYI there are some ongoing efforts to extend Marlin to support W4A16 and W8A16

Right now the kernels load GPTQ models, but we could really connect them to any models type

We should run benchmarks against these as well in deciding which kernel to use

https://github.com/vllm-project/vllm/pull/7079

**Comment by mzusman** [comment]
> @mzusman
> 
> How to convert models to ExpertsInt8?

You would just need to run vLLM with `--quantization experts_int8`, It supports quantization on the fly

**Comment by mzusman** [comment]
> Hey - just FYI there are some ongoing efforts to extend Marlin to support W4A16 and W8A16
> 
> Right now the kernels load GPTQ models, but we could really connect them to any models type
> 
> We should run benchmarks against these as well in deciding which kernel to use
> 
> #7079

I understand, I was trying to benchmark this method again the PR #7079  on `Nous-Hermes-2-Mixtral-8x7B-DPO-GPTQ-8bit-128g`/'TheBloke/Mixtral-8x7B-v0.1-GPTQ' and it ended up unsuccessful (hanging on startup).. 

Do you think this is a blocker for merging this PR? We can have two options available.

**Comment by jeejeelee** [comment]
It seems that #6502 is also a similar PR.

File: tests/models/test_experts_int8.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# flake8: noqa
```
Comment (mgoin) [inline tests/models/test_experts_int8.py:1]: Maybe this should be moved to `tests/quantization/test_experts_int8.py`

File: tests/models/test_experts_int8.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+"""Tests experts_int8 quantization startup and generation, 
+doesn't test correctness
+"""
+
+import pytest
+
+MODELS = ["ai21labs/Jamba-tiny-random"]
+
+
+@pytest.mark.parametrize("model", MODELS)
```
Comment (mgoin) [inline tests/models/test_experts_int8.py:11]: This should be disabled on GPUs that can't support the quant method
```python
from tests.quantization.utils import is_quant_method_supported

@pytest.mark.skipif(not is_quant_method_supported("experts_int8"),
                    reason="ExpertsInt8 is not supported on this GPU type.")
```

File: benchmarks/kernels/benchmark_moe.py [hunk 3]
```diff
@@ -224,20 +249,15 @@ def sort_config(config: BenchmarkConfig) -> BenchmarkConfig:
...
-    hidden_size: int,
-    topk: int,
-    dtype: torch.dtype,
-    use_fp8: bool,
-) -> None:
-    dtype_str = "float8" if use_fp8 else None
+def save_configs(configs: Dict[int, BenchmarkConfig], num_experts: int,
+                 shard_intermediate_size: int, hidden_size: int, topk: int,
+                 dtype: torch.dtype, use_fp8: bool, use_int8: bool) -> None:
+    dtype_str = "float8" if use_fp8 else "int8" if use_int8 else None
```
Comment (mgoin) [inline benchmarks/kernels/benchmark_moe.py:255]: Let's break this out, stacked ternaries are a bit much. Also consider the rename I posted above
```suggestion
    dtype_str = None
    if use_fp8:
        dtype_str = "float8"
    elif use_int8:
        dtype_str = "int8"
```

File: benchmarks/kernels/benchmark_moe.py [hunk 4]
```diff
@@ -31,18 +31,35 @@ def benchmark_config(
     topk: int,
     dtype: torch.dtype,
     use_fp8: bool,
+    use_int8: bool,
```
Comment (mgoin) [inline benchmarks/kernels/benchmark_moe.py:34]: `use_fp8` benchmarks FP8 with W8A8 while your current implementation of INT8 is just W8A16. Given this I would prefer to be more explicit here so could you please change them to `use_fp8_w8a8` and `use_int8_w8a16` respectively?

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 5]
```diff
@@ -17,42 +17,44 @@
...
+        # Meta-parameters
+        BLOCK_SIZE_M: tl.constexpr,
+        BLOCK_SIZE_N: tl.constexpr,
+        BLOCK_SIZE_K: tl.constexpr,
+        GROUP_SIZE_M: tl.constexpr,
+        MUL_ROUTED_WEIGHT: tl.constexpr,
+        top_k: tl.constexpr,
+        compute_type: tl.constexpr,
+        use_fp8: tl.constexpr,
+        use_int8: tl.constexpr):
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:57]: Again, especially in the kernel, I think it is really important to distinguish that the `use_fp8` path uses quantized weights and activations, while the `use_int8` path uses only quantized weights. In the future we will want to add INT8 W8A8, so it is important to leave space for that

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 6]
```diff
@@ -229,13 +238,15 @@ def invoke_fused_moe_kernel(A: torch.Tensor, B: torch.Tensor, C: torch.Tensor,
...
     assert topk_weights.stride(1) == 1
     assert sorted_token_ids.stride(0) == 1
 
-    if not use_fp8:
+    if not use_fp8 and not use_int8:
         assert A_scale is None
         assert B_scale is None
+    elif use_int8:
+        assert B_scale is not None
     else:
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:250]: Maybe it is best to reverse this ordering so `if not use_fp8 and not use_int8` just becomes the `else` case

**Comment by qingquansong** [comment]
> I tested this pull request on deepseek-v2-chat-236B. Indeed more concurrency.

Hey @halexan thanks for the great efforts！I have a previous pr for w8a8  here https://github.com/vllm-project/vllm/pull/6978 would it be possible to merge the two efforts into one
pr to yours and optimize together to check in？Thank you！

Best regards，
QQ

File: vllm/model_executor/layers/quantization/experts_int8.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    def get_name(cls) -> str:
+        return "experts_int8"
+
+    @classmethod
+    def get_supported_act_dtypes(cls) -> List[torch.dtype]:
+        return [torch.bfloat16, torch.half]
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        return 80
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/experts_int8.py:30]: Do you know that older GPUs won't work here?

File: vllm/model_executor/layers/quantization/experts_int8.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+                       hidden_size: int, intermediate_size: int,
+                       params_dtype: torch.dtype, **extra_weight_attrs):
+
+        int8_dtype = torch.int8
+
+        assert 'weight_loader' in extra_weight_attrs
+        weight_loader = extra_weight_attrs['weight_loader']
+        wrapped_weight_loader = ExpertsInt8MoEMethod.quantizing_weight_loader(
+            layer, weight_loader)
+        extra_weight_attrs['weight_loader'] = wrapped_weight_loader
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/experts_int8.py:67]: Does this take care of quantizing the weights as each one is loaded? We should consider this idea for other dynamic quant methods like fp8

**Review by mgoin — COMMENTED** [review]
This generally looks good to me and experiments with some interesting ideas. I think the biggest issue is the clashing assumption between `use_fp8` and `use_int8`, where one is W8A8 and the other is W8A16 - we should definitely be explicit with the level of quantization if we're going this route.

It would be nice if we could select between INT8 W8A16 or W8A8 (since we already have efficient activation quant methods), like @qingquansong has proposed in https://github.com/vllm-project/vllm/pull/6978

In the short-term it seems like we might end up with `use_fp8_w8a8`, `use_int8_w8a8`, and `use_int8_w8a16` - so it would be nice if we could work towards a more sustainable future for the code.

File: vllm/model_executor/layers/quantization/experts_int8.py (hunk 8 above)
Comment (mzusman) [reply vllm/model_executor/layers/quantization/experts_int8.py:67]: That's right, wrapping the weight loader allows us to immediately quantize the weight on the fly in-place just before it is copied onto the state dict, we also run the quantization process on device so it's even faster.

File: vllm/model_executor/layers/quantization/experts_int8.py (hunk 7 above)
Comment (mzusman) [reply vllm/model_executor/layers/quantization/experts_int8.py:30]: We tried Volta (V100) and it didn't work. Didn't try to run on T4.

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 6 above)
Comment (mzusman) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:250]: Addressed :)

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 5 above)
Comment (mzusman) [reply vllm/model_executor/layers/fused_moe/fused_moe.py]: Addressed.
I've changed the config_dtype string to int8_w8a16 as well, for future accommodation of the w8a8 moe config files and other methodologies.

File: benchmarks/kernels/benchmark_moe.py (hunk 4 above)
Comment (mzusman) [reply benchmarks/kernels/benchmark_moe.py]: Addressed

File: benchmarks/kernels/benchmark_moe.py (hunk 3 above)
Comment (mzusman) [reply benchmarks/kernels/benchmark_moe.py]: Thank you, Addressed

File: .buildkite/run-cpu-test.sh [hunk 9]
```diff
@@ -23,7 +23,7 @@ docker exec cpu-test-avx2 bash -c "python3 examples/offline_inference.py"
 # Run basic model test
 docker exec cpu-test bash -c "
   pip install pytest Pillow protobuf
-  pytest -v -s tests/models -m \"not vlm\" --ignore=tests/models/test_embedding.py --ignore=tests/models/test_oot_registration.py --ignore=tests/models/test_registry.py --ignore=tests/models/test_jamba.py --ignore=tests/models/test_danube3_4b.py" # Mamba and Danube3-4B on CPU is not supported
+  pytest -v -s tests/models -m \"not vlm\" --ignore=tests/models/test_embedding.py --ignore=tests/models/test_oot_registration.py --ignore=tests/models/test_registry.py --ignore=tests/models/test_jamba.py --ignore=tests/models/test_danube3_4b.py --ignore=tests/models/test_experts_int8.py" # Mamba, ExpertsInt8 and Danube3-4B on CPU is not supported
```
Comment (mgoin) [inline .buildkite/run-cpu-test.sh:26]: I think you can remove this change now

File: tests/models/test_experts_int8.py (hunk 1 above)
Comment (mzusman) [reply tests/models/test_experts_int8.py]: Moved it to quantization sub dir

File: tests/models/test_experts_int8.py (hunk 2 above)
Comment (mzusman) [reply tests/models/test_experts_int8.py]: Addressed

File: benchmarks/kernels/benchmark_moe.py [hunk 10]
```diff
@@ -262,7 +292,8 @@ def main(args: argparse.Namespace):
 
     hidden_size = config.hidden_size
     dtype = config.torch_dtype
-    use_fp8 = args.dtype == "fp8"
+    use_fp8_w8a8 = args.dtype == "fp8"
```
Comment (mgoin) [inline benchmarks/kernels/benchmark_moe.py:297]:
```suggestion
    use_fp8_w8a8 = args.dtype == "fp8_w8a8"
```

File: benchmarks/kernels/benchmark_moe.py [hunk 11]
```diff
@@ -323,7 +354,7 @@ def _distribute(method: str, inputs: List[Any]) -> List[Any]:
     parser.add_argument("--tp-size", "-tp", type=int, default=2)
     parser.add_argument("--dtype",
                         type=str,
-                        choices=["auto", "fp8"],
+                        choices=["auto", "fp8", "int8"],
```
Comment (mgoin) [inline benchmarks/kernels/benchmark_moe.py:359]:
```suggestion
                        choices=["auto", "fp8_w8a8", "int8_w8a16"],
```

File: benchmarks/kernels/benchmark_moe.py [hunk 12]
```diff
@@ -155,11 +177,16 @@ def benchmark(
...
+        use_int8_w8a16: bool,
     ) -> Tuple[Dict[str, int], float]:
         torch.cuda.manual_seed_all(self.seed)
 
-        dtype_str = "float8" if use_fp8 else None
+        dtype_str = None
+        if use_fp8_w8a8:
+            dtype_str = "float8"
+        elif use_int8_w8a16:
+            dtype_str = "int8"
```
Comment (mgoin) [inline benchmarks/kernels/benchmark_moe.py:189]: Should these strings be make explicit as well, or are they used as torch dtypes?

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 13]
```diff
@@ -426,14 +440,29 @@ def grouped_topk(hidden_states: torch.Tensor,
     return topk_weights, topk_ids
 
 
+def get_config_dtype_str(dtype: torch.dtype,
+                         use_int8_w8a16: Optional[bool] = False,
+                         use_fp8_w8a8: Optional[bool] = False):
+    if use_fp8_w8a8:
+        return "fp8"
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:447]:
```suggestion
        return "fp8_w8a8"
```

File: vllm/model_executor/layers/quantization/experts_int8.py (hunk 7 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/experts_int8.py:30]: Okay that is fair enough, thanks

**Review by mgoin — COMMENTED** [review]
Thanks for the quick changes! I think these are my last round of comments

**Comment by mzusman** [comment]
> This generally looks good to me and experiments with some interesting ideas. I think the biggest issue is the clashing assumption between `use_fp8` and `use_int8`, where one is W8A8 and the other is W8A16 - we should definitely be explicit with the level of quantization if we're going this route.
> 
> It would be nice if we could select between INT8 W8A16 or W8A8 (since we already have efficient activation quant methods), like @qingquansong has proposed in #6978
> 
> In the short-term it seems like we might end up with `use_fp8_w8a8`, `use_int8_w8a8`, and `use_int8_w8a16` - so it would be nice if we could work towards a more sustainable future for the code.

Thank you for the thorough review! I've changed the terms `use_int8`/`use_fp8` to `use_int8_w8a16`/`use_fp8_w8a8` and delivered this idea to the moe config files dtype as well.

File: vllm/model_executor/layers/fused_moe/fused_moe.py (hunk 13 above)
Comment (mzusman) [reply vllm/model_executor/layers/fused_moe/fused_moe.py]: I'll change the moe config filename as well.

File: benchmarks/kernels/benchmark_moe.py (hunk 12 above)
Comment (mzusman) [reply benchmarks/kernels/benchmark_moe.py]: They are used to load the moe config files, changed it to use the new function that encapsulates this logic in one place

File: benchmarks/kernels/benchmark_moe.py (hunk 11 above)
Comment (mzusman) [reply benchmarks/kernels/benchmark_moe.py]: Addressed

File: benchmarks/kernels/benchmark_moe.py (hunk 10 above)
Comment (mzusman) [reply benchmarks/kernels/benchmark_moe.py]: Addressed

File: .buildkite/run-cpu-test.sh (hunk 9 above)
Comment (mzusman) [reply .buildkite/run-cpu-test.sh]: Addressed

**Comment by mzusman** [comment]
CI failures seem to not be related to this PR

File: vllm/model_executor/layers/quantization/experts_int8.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+                                     shard_size].copy_(scales[:, 0])
+            # w2, down_proj case: Load into only shard of w2.
+            elif shard_id == "w2":
+                scales = quantize_in_place_and_get_scales(loaded_weight[:,
+                                                                        shard])
+                layer.w2_scale.data[expert_id, :].copy_(scales[:, 0])
+            else:
+                raise ValueError(
+                    f"Shard id must be in [0,1,2] but got {shard_id}")
+            weight_loader(param, loaded_weight, weight_name, shard_id,
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/experts_int8.py:161]: Just an fyi - we're updating/expanding the weight loading for Fused MoE layers: https://github.com/vllm-project/vllm/pull/7527

**Review by mgoin — APPROVED** [review]
LGTM! Ditto with Dipika that it'd be good to make this work with the MoE Parameter refactor eventually

**Comment by mzusman** [comment]
Thanks! I'll rebase, maybe will resolve the CI issues

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 15]
```diff
@@ -588,7 +623,9 @@ def fused_moe(
     - topk_group: Optional[int]: additional parameter for grouped_topk
     - use_grouped_topk: If True, use grouped_topk instead of fused_topk
         note: Deepseekv2 model uses grouped_topk
-    - use_fp8 (bool): If True, use fp8 arithmetic to compute the inner
+    - use_fp8_w8a8 (bool): If True, use fp8 arithmetic to compute the inner
+        products for w1 and w2. Defaults to False.
+    - use_int8_w8a16 (bool): If True, use fp8 arithmetic to compute the inner
```
Comment (AllenDou) [inline vllm/model_executor/layers/fused_moe/fused_moe.py:628]: fp8?

**Pull request merged**